### ISSUE TITLE
Mechwarrior 1989 MT32 music slowdown

### ISSUE THREAD

**Issue by zzgundamx** [joncampbell123/dosbox-x#2765]
Dosbox-X latest version 0.83.16 when I run Mechwarrior 1989 through it the music plays at a very slow tempo (MT32). I had to use the CPU speed up F11+] to increase the speed until it plays at the correct tempo. It affects all music in game (what little there is), at the intro and at the bar.

This behaviour is observed both when using Dosbox-X's built in MT32 emulation and when using MUNT separately (MUNT 2.5.0).

Strangely enough all other MT32 games I have tested is unaffected - Battletech 2, Ultima VI and Hero's Quest.

**Comment by zzgundamx** [comment]
On further testing, I discovered that the music plays at the correct temp only at 1000 cycles. Any higher will induce the slowdown. However, changing the CPU cycle mid-game will cause the keyboard to scroll up continously - another bug.

**Comment by Allofich** [comment]
> However, changing the CPU cycle mid-game will cause the keyboard to scroll up continously - another bug.

That might not be a bug. When you changed the CPU cycles mid-game, that might have confused the game if it stored timing-related values based on the old speed.

**Comment by zzgundamx** [comment]
Thanks Allofich. Do you have any suggestion as to the MT32 music slowdown?

**Comment by Allofich** [comment]
Not really, except that generally speaking it doesn't seem to be uncommon that old games have a "sweet spot" of cycles, away from which they run too fast, too slow or sometimes with problems like graphical glitches.

My feeling, though, is that we need to determine whether there is for sure a problem here in DOSBox-X or not. (Maybe from your view or your familiarity with the game you are sure DOSBox-X's behavior is mistaken, but it's not obvious to me.) Could it just be that Mechwarrior's music tempo behaves like that at different CPU speeds?

The game's box
https://www.mobygames.com/game/mechwarrior/cover-art/gameCoverId,114/
says "8MHz recommended". DOSBox-X's "Emulate CPU speed" menu has an 8MHz option which puts the emulator at 750 cycles. How does the tempo sound at 750?

> Dosbox-X latest version 0.83.16 when I run Mechwarrior 1989 through it the music plays at a very slow tempo (MT32). I had to use the CPU speed up F11+] to increase the speed until it plays at the correct tempo.

What was the baseline where the temp was too slow? Was the rest of the game (both 2D and 3D parts) running at a correct speed, with only the music being too slow?

> I discovered that the music plays at the correct tempo only at 1000 cycles. Any higher will induce the slowdown.

Does this mean the tempo was too slow below 1000, correct at 1000, then too slow again above 1000? 

Another thing that can always potentially be helpful is knowing whether the same thing happens in original DOSBox, if you feel like testing that.

**Comment by zzgundamx** [comment]
> Does this mean the tempo was too slow below 1000, correct at 1000, then too slow again above 1000?
> 
> Another thing that can always potentially be helpful is knowing whether the same thing happens in original DOSBox, if you feel like testing that.

Yes correct, the tempo is too slow below "and" again above 1000 as you have stated. I have actually tested Dosbox ECE on the same PC it exhibited the same behaviour. However, on my other PC with Dosbox ECE it plays normally! 

I've also got my friend to test Dosbox-X (I sent him my config and same Dosbox-X 0.83.16 and he does not have an issue which is really strange.

Thanks for the input so far - this is a really strange outcome and Mechwarrior 1989's MT32 is the only one that have exhibited this issue. It's Adlib music though plays just fine.

**Comment by Allofich** [comment]
Don't know.. all I can suggest is trying to re-install things on the problem PC or try posting at the VOGONS forums.

**Comment by grapeli** [comment]
Four different versions, which one is closer to your pattern (bar theme).
DOSBox-X, different cycles settings in two cases, external `mt32d` used.
[01.flac.txt](https://github.com/joncampbell123/dosbox-x/files/6962380/01.flac.txt)
[02.flac.txt](https://github.com/joncampbell123/dosbox-x/files/6962381/02.flac.txt)
[03.flac.txt](https://github.com/joncampbell123/dosbox-x/files/6962383/03.flac.txt)
[04.flac.txt](https://github.com/joncampbell123/dosbox-x/files/6962384/04.flac.txt)

**Comment by grapeli** [comment]
Another version from dosbox-staging.The duration is identical (57.433), but the sound is different. It seems much closer to the version found on [youtube](https://www.youtube.com/watch?v=oyNxxeGab2I).
[06.flac.txt](https://github.com/joncampbell123/dosbox-x/files/6964411/06.flac.txt)
`cycles=10000`

**Comment by zzgundamx** [comment]
the 06.flac is the closest to what I have grapeli. So it seems in your testing, there is no slowdown of any kind encountered.

i'm using the latest Munt 2.5.1 - MT32 emulator.

**Comment by grapeli** [comment]
@zzgundamx 
Are you using an external mt32? For what reason.

DOSBox-X uses the supplied [libmt32](https://github.com/joncampbell123/dosbox-x/tree/master/src/libs/mt32) library. It is statically linked to the dosbox-x binary. Its current version is [2.5.0](https://github.com/joncampbell123/dosbox-x/blob/945905bb7b3e56bbc3e6be1b62c1fd73aaa11f74/src/libs/mt32/Synth.h#L21) (last stable release).

The sound differences are either a matter of different settings or different implementations in the code.
Dosbox-staging uses [mt32emu's floating point synthesis](https://github.com/dosbox-staging/dosbox-staging/commit/ff6b00a70e9a10062963f07104803d09b89fd6bf).

**Comment by zzgundamx** [comment]
hey grapeli - no as stated I'm using the Munt MT32 emulator (ver 2.5.1).

You mentioned the libmt32 library 2.5.0 - is the current dosbox-x 0.83.16 using this version as well or is there a newer build that I can download.

**Comment by rderooy** [comment]
For the Linux packages that I'm building (flatpak and RPM) I'm using or linking to 2.5.2

**Comment by grapeli** [comment]
I was wrong last libmt32 release was 2.5.3 (6 days ago).

@rderooy You have created a non-existent (empty) dependency.

**Comment by rderooy** [comment]
I have? where?

**Comment by grapeli** [comment]
Update to version 2.5.3 is not difficult. I am not a programmer, I do not know if this is fully correct. Maybe I missed something (probably yes). I did it only as a test to compare the sound in Mechwarrior. It is identical to version 2.5.0.

After updating the code, I made such a change.
```
find src/libs/mt32 -type f \( -name "*.cpp" -o -name "*.h" \) -print0 | xargs -0 sed -i 's/Bit32u/uint32_t/g;s/Bit32s/int32_t/g;s/Bit16u/uint16_t/g;s/Bit8u/uint8_t/g;s/Bit8s/int8_t/g'    
sed -i 's/renderint16_t/renderBit16s/g' src/gui/midi_mt32.h
sed -i '/#define MT32EMU_SYNTH_H/a #define MT32EMU_VERSION "2.5.0"\n#define MT32EMU_VERSION_MAJOR 2\n#define MT32EMU_VERSION_MINOR 5\n#define MT32EMU_VERSION_PATCH 3' libs/mt32/Synth.h src/libs/mt32/Synth.h
```

**Comment by grapeli** [comment]
Unfortunately, the mt32 sound from dosbox-x in Mechwarrior (1989) differs slightly from the original and from dosbox-staging. I have tried many settings without a satisfactory result.

[Intro theme](https://www.youtube.com/watch?v=oLfCTMoumSs) (orginal mt32).
[orginal.mt32.flac.txt](https://github.com/joncampbell123/dosbox-x/files/6988710/orginal.mt32.flac.txt)
Dosbox-staging offers a very similar quality to the original mt32.
[dosbox-staging.mt32.flac.txt](https://github.com/joncampbell123/dosbox-x/files/6988711/dosbox-staging.mt32.flac.txt)
Dosbox-x is a little different.
[dosbox-x.mt32.flac.txt](https://github.com/joncampbell123/dosbox-x/files/6988712/dosbox-x.mt32.flac.txt)

Difference it is clearly visible on a very short sample.
[sample.orginal.flac.txt](https://github.com/joncampbell123/dosbox-x/files/6988714/sample.orginal.flac.txt)
[sample.dosbox-x.flac.txt](https://github.com/joncampbell123/dosbox-x/files/6988715/sample.dosbox-x.flac.txt)

After generating the spectrogram, we see that the key frequencies are lowered compared to the original sound. Dosbox-x lowers the tone.
```
ffmpeg -v quiet -i sample.orginal.flac  -lavfi showspectrumpic=s=8192x4096:orientation=horizontal:start=250:stop=7000 spect.orginal.png
ffmpeg -v quiet -i sample.dosbox-x.flac -lavfi showspectrumpic=s=8192x4096:orientation=horizontal:start=250:stop=7000 spect.dosbox-x.png
```
orginal: 585,1180,1770,2360,2952,3545,4140,...
dosbox-x: 540,1090,1620,2175,2715,3265,3785,...

You can slightly raise the tone and you will get a very similar effect to the original.
https://www.breakfastquay.com/rubberband/
`rubberband -P -p 1.46 sample.dosbox-x.wav new.sample.dosbox-x.wav`

**Comment by maron2000** [comment]
Hi, I also found that the pitch of music played from built-in MT32 emulation entirely a bit low compared to when using external MUNT emulator (2.5.0).
Tested on 0.83.16 Windows 64bit SDL1.

**Comment by grapeli** [comment]
@maron2000 
On linux with dosbox-x using external mt32d (mt32emu_alsadrv) the pitch is identical to the [youtube sample](https://www.youtube.com/watch?v=oLfCTMoumSs).

original mt32
![original mt32](https://user-images.githubusercontent.com/452325/129616157-cbe16ab1-ba08-4786-ad71-c86b7d1f4208.png)
dosbox-x external mt32
![dosbox-x external mt32](https://user-images.githubusercontent.com/452325/129616164-e40e3ae3-aff2-4b3b-a1a4-a895daf2c38a.png)
dosbox-x built-in mt32
![dosbox-x built-in mt32](https://user-images.githubusercontent.com/452325/129616176-8aa8292a-ebc8-4160-97d2-9b222e8d42e9.png)

**Comment by rderooy** [comment]
One of the dosbox-staging developers, @kcgen, did a lot of cleanup to the -staging sound code, including MT32 which would probably explain the difference.

Unfortunately, due to divergence of -staging with the main dosbox tree, porting these changes will probably not be trivial.

**Comment by maron2000** [comment]
@rderooy I do respect the efforts to integrate MT32 emulation, but may be difficult to fix bugs if they arise.

@grapeli  I noticed that the sampling frequency shown in each of the 3 figures you have shown, built-in MT-32 is 44.1kHz while original and external is 48kHz. Will this be any clue for the fix?

**Comment by grapeli** [comment]
@maron2000 
Of course. That explains it all.
With the sample rate of 48000, the pitch is the same for the built-in mt32.
dosbox-x built-in mt32.rate=48000
![dosbox-x built-in mt32 rate 48000](https://user-images.githubusercontent.com/452325/129697163-a925a69c-f2e0-4cf9-bc48-ac68257d54e2.png)

In dosbox-staging I had the sampling rate set to 44100 and the pitch was the same as for this recording (from YouTube).

**Comment by maron2000** [comment]
@grapeli I wasn't aware of `mt32.rate` option.
I also confirmed that the pitch of music is as same as external emulation.
I think the pitch shouldn't change by changing this option, but at least we found a workaround, it's nice.

**Comment by grapeli** [comment]
@maron2000 
https://github.com/joncampbell123/dosbox-x/blob/master/src/libs/mt32/Synth.cpp#L230

The internal sampling rate for mt32 is 32kHz. It is defined [here](https://github.com/joncampbell123/dosbox-x/blob/master/src/libs/mt32/globals.h#L73).
For [getStereoOutputSampleRate](https://github.com/joncampbell123/dosbox-x/blob/master/src/libs/mt32/Synth.cpp#L230) (mt32.analog) it is:
0 = 32000
1 = 32000
2 = 48000 **default** (32000 * 3/2)
3 = 96000 (32000 * 3)

Dosbox-staging no separate sampling rate settings for mt32, only globally for mixer.

**Comment by maron2000** [comment]
@grapeli I see. I think I sort of understood your explanation.
Thanks for the information!
